tt_eager/tt_dnn/op_library/run_operation.hpp

```diff
@@ -401,4 +402,256 @@ std::vector<Device*> get_workers_for_op_output(const std::vector<Tensor>&& input
 
 } //namespace operation
 
+bool validate_worker_modes(const std::vector<Device*>& workers);
+
+class TensorMonad {
```

Comment:
Is TensorMonad required?

We can just run the entire composite in the worker thread, as originally planned.

Reply:
The TensorMonad allows you to have separate ops that can run in the same work thread.   This is following the same pattern as if we called multiple launch_ops back to back.   The difference is that we are formally identifying the parts of the code that validates the api, that creates the output tensors, and that creates the next op after the prior op has run.

We formally define within the op class itslef.
static void validate_api_arguments(...);
static std::vector<std::optional<Tensor>> create_async_output_tensors(...);
static Reshard create_async_operation(...);

In the current design these three aspects of the op lives outside of the op itself.   
Within the TensorMonad.bind<OpName> we do chain the launch_op calls via the equivalent launch_async.

In the end this is the exact same pattern that was already planned but this just simply tries to put all these aspects into their proper place.

Reply:
> The TensorMonad allows you to have separate ops that can run in the same work thread. This is following the same pattern as if we called multiple launch_ops back to back. The difference is that we are formally identifying the parts of the code that validates the api, that creates the output tensors, and that creates the next op after the prior op has run.
> 
> We formally define within the op class itslef. static void validate_api_arguments(...); static std::vector<std::optional> create_async_output_tensors(...); static Reshard create_async_operation(...);
> 
> In the current design these three aspects of the op lives outside of the op itself. Within the TensorMonad.bind we do chain the launch_op calls via the equivalent launch_async.
> 
> In the end this is the exact same pattern that was already planned but this just simply tries to put all these aspects into their proper place.

Any number of OPs can run in the worker thread as is, worker thread runs then sequentially and the semantics within the worker thread are "full synchronous". 

An example from this commit : https://github.com/tenstorrent/tt-metal/pull/8139 , shows it is straightforward to run multiple OPs in the worker thread:

```cpp
Tensor _mish(const Tensor& x, const MemoryConfig& output_mem_config) {
    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({x}))};
    operation::launch_op(
        [output_mem_config] (const std::vector<Tensor>& input_tensors, const std::vector<std::optional<const Tensor>>& optional_input_tensors) mutable -> std::vector<Tensor> {
            const auto& x = input_tensors.at(0);
            Tensor sp_x = softplus(x, 1.0f, 20.0f, output_mem_config);
            Tensor tanh_x = tanh(sp_x, output_mem_config);
            sp_x.deallocate();
            Tensor mish_x = mul(x, tanh_x, std::nullopt, output_mem_config);
            return {mish_x};
        }, {x}, output_tensors);
    return output_tensors.at(0);
}
```